CRM/Civicase/Service/CaseTypeCategoryEventHandler.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+/**
+ * Class CRM_Civicase_Service_CaseTypeCategoryEventHandler.
+ */
+class CRM_Civicase_Service_CaseTypeCategoryEventHandler {
+
+  /**
+   * Menu handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryMenu
+   */
+  protected $menu;
+
+  /**
+   * Custom data handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryCustomDataType
+   */
+  protected $customData;
+
+  /**
+   * Custom field handler.
+   *
+   * @var \CRM_Civicase_Service_CaseCategoryCustomFieldExtends
+   */
+  protected $customFieldExtends;
+
+  /**
+   * Should we check referer before running action?
+   *
+   * @var bool
+   */
+  protected $needCheckReferer = FALSE;
+
+  /**
+   * Should we check option group before running action?
+   *
+   * Depends on $optionGroupId.
+   *
+   * @var bool
+   */
+  protected $needCheckOptionGroup = TRUE;
+
+  /**
+   * Case type category option group id. Used to check option group.
+   *
+   * @var int
+   */
+  protected $optionGroupId;
+
+  /**
+   * Object name.
+   *
+   * Is used to check object if it's type is unpredictable or unknown.
+   *
+   * @var string
+   */
+  protected $objectName = 'OptionValue';
+
+  /**
+   * CRM_Civicase_Service_CaseTypeCategoryEventHandler constructor.
+   *
+   * @param object $menu
+   *   Menu handler.
+   * @param object $customData
+   *   Custom data handler.
+   * @param object $customFieldExtends
+   *   Custom field handler.
+   */
+  public function __construct($menu, $customData, $customFieldExtends) {
+    $this->menu = $menu;
+    $this->customData = $customData;
+    $this->customFieldExtends = $customFieldExtends;
+  }
+
+  /**
+   * Perform actions on case type category create.
+   *
+   * @param string $name
+   *   Case type category name.
+   */
+  public function onCreate($name) {
+    if (!$name || !$this->shouldRun()) {
+      return;
+    }
+
+    $this->menu->createItems($name);
+    $this->customFieldExtends->create($name, "Case ({$name})");
+    $this->customData->create($name);
+  }
+
+  /**
+   * Perform actions on case type category create.
+   *
+   * @param array $category
+   *   Case type category data, should contain:
+   *   id - category id,
+   *   is_active - is_active category property (enabled/disabled),
+   *   icon - category icon.
+   */
+  public function onUpdate(array $category) {
+    if (empty($category['id']) || !$this->shouldRun()) {
+      return;
+    }
+
+    $updateParams = [];
+    if (isset($category['is_active'])) {
+      $updateParams['is_active'] = !empty($category['is_active']) ? 1 : 0;
+    }
+    if (isset($category['icon'])) {
+      $updateParams['icon'] = 'crm-i ' . $category['icon'];
+    }
+
+    if (!empty($updateParams)) {
+      $this->menu->updateItems($category['id'], $updateParams);
+    }
+  }
+
+  /**
+   * Removes case type category menu item from the civicrm navigation bar.
+   *
+   * @param string $name
```

Comment:
Can we rename this to caseCategoryName.

Reply:
All this class is about case type categories. So why should we add caseCategory prefix to all variables? Why not follow 'Write less - Do more' approach in this case?

Reply:
I responded to this in an earlier comment